src/charm.py

```diff
@@ -291,6 +303,98 @@ def _on_db_router_relation_changed(self, event: RelationChangedEvent) -> None:
 
         self.unit.status = ActiveStatus()
 
+    def _on_shared_db_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy shared_db relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        if not self.unit.is_leader():
+            return
+
+        self.unit.status = MaintenanceStatus("Setting up shared-db relation")
+        logger.warning("DEPRECATION WARNING - `shared-db` is a legacy interface")
+
+        unit_relation_databag = event.relation.data[self.unit]
+        app_relation_databag = event.relation.data[self.app]
+
+        if unit_relation_databag.get("password"):
+            # Test if relation data is already set
+            # and avoid overwriting it
+            logger.warning("Data for shared-db already set.")
+            self.unit.status = ActiveStatus()
+            return
+
+        # retrieve data from the relation databag
+        # Abstracted is the fact that it is received from the leader unit of the related app
+        requires_relation_databag = event.relation.data[event.unit]
+        database_name = requires_relation_databag.get("database")
+        database_user = requires_relation_databag.get("username")
+
+        if not database_name or not database_user:
+            # Cannot create scoped database without credentials
+            logger.warning(
+                "Missing information for shared-db relation to create a database and scoped user"
+            )
+            self.unit.status = ActiveStatus()
+            return
+
+        password = generate_random_password(PASSWORD_LENGTH)
+
+        try:
+            self._mysql.create_application_database_and_scoped_user(
+                database_name, database_user, password
+            )
+
+            # set the relation data for consumption
+            cluster_primary = str(self.model.get_binding(PEER).network.bind_address)
+
+            unit_relation_databag["db_host"] = cluster_primary
+            # Database port is static in legacy charm
+            unit_relation_databag["db_port"] = "3306"
+            # Wait timeout is a config option in legacy charm
+            # defaulted to 3600 seconds
+            unit_relation_databag["wait_timeout"] = "3600"
+            unit_relation_databag["password"] = password
+
+            unit_names = " ".join([unit.name for unit in event.relation.units])
+            unit_relation_databag["allowed_units"] = unit_names
```

Comment:
im wondering if we should be updating `allowed_units` with every unit that joins the relation (and not only the first time the database user and application database are created)

Reply:
now that i think of it, we may also need to update `allowed_units` upon the relation departed event (when units leave the relation)

Reply:
You are right, but bear in mind this is mainly a backward compatibility thing, since this data is not being used. IIRC the reactive charm did not allow connections from every host (e.g. `bind 0.0.0.0`). Since it's not the case here, this does not really matter - will add the relation departing part though.